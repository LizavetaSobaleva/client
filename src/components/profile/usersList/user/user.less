@import '../../../../utils/colors.less';

@import '../../../../utils/colors.less';

.user {
    padding: 8px 12px;
    border-bottom: 1px solid @lightgrey;
    display: grid;
    grid-template-columns: 1.5fr 2fr 2fr 2fr 0.5fr;
    color: @subtext;
    font-weight: 400;
    align-items: center;
    transition: ease-in 50ms all;

    &:hover {
        background-color: #f9fafc;
    }

    &__name {
        grid-column-start: 1;
        width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__email {
        grid-column-start: 2;
        width: 98%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }

    &__status {
        grid-column-start: 3;
        justify-self: center;

        select {
            padding: 4px 16px;
            border: none;
            border-radius: 16px;
            background-color: #f9fafc;
            color: @primary;
            font-size: 16px;
            cursor: pointer;
        
            &:focus {
              outline: none;
            }
          }
    }

    &__size {
        grid-column-start: 4;
        width: auto;
        display: flex;
        flex-direction: column;
        font-size: 12px;
        padding: 4px 20px;

        .progressBar {
            width: auto;
            padding: 4px 0;

            &__line, &__upload {
                height: 8px;
            }

            &__percent {
                display: none;
            }
        }
    }

    &__actions {
        grid-column-start: 5;

        .primaryButton {
            padding: 8px 24px;
        }
    }
}

@media only screen and (max-width: 660px){
    .user {
        grid-template-columns: 1.5fr 2fr 2fr;

        &__name {
            margin-left: 4px;
        }

        &__email {
            display: none
        }

        &__status {
            grid-column-start: 2;
        }
    
        &__size {
            grid-column-start: 3;
        }
    }
}