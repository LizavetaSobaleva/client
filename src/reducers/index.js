import { applyMiddleware, combineReducers } from "redux";
import { configureStore} from "@reduxjs/toolkit"
import { thunk } from "redux-thunk";
import userReducer from "./userReducer";
import fileReducer from "./fileReducer";
import uploadReducer from "./uploadReducer";
import appReducer from "./appReducer";
import { composeWithDevTools } from '@redux-devtools/extension';


const rootReducer = combineReducers({
    user: userReducer,
    files: fileReducer,
    upload: uploadReducer,
    app: appReducer
})

export const store = configureStore({reducer: rootReducer, devTools: composeWithDevTools(applyMiddleware(thunk))})