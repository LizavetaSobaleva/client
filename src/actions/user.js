import axios from "axios"
import { setUser } from '../reducers/userReducer'


export const instanceAxios = axios.create({
     baseURL: process.env.REACT_APP_API_URL + "/api/"
})

export const registration = async (name, email, password, dispatch) => {
   try {
        const response = await instanceAxios.post('auth/registration', {
            name,
            email,
            password
        })
        alert(response.data.message)
        dispatch(login(email, password))
   } catch (e) {
        alert(e.response.data.message)
   }
}

export const login = (email, password) => {
     return async dispatch => {
          try {
               const response = await instanceAxios.post('auth/login', {
                   email,
                   password
               })
               dispatch(setUser(response.data.user))
               localStorage.setItem('token', response.data.token)
          } catch (e) {
               alert(e.response.data.message)
          }
     }
}

export const auth = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.get('auth/auth',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               localStorage.setItem('token', response.data.token)
               dispatch(setUser(response.data.user))
          } catch (e) {
               console.log(e.response.data.message)
               localStorage.removeItem('token')
          }
     }
}

export const uploadAvatar = (file) => {
     return async dispatch => {
          try {
               const formData = new FormData()
               formData.append('file', file)
               const response = await instanceAxios.post('files/avatar', formData,
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}

export const deleteAvatar = () => {
     return async dispatch => {
          try {
               const response = await instanceAxios.delete('files/avatar',
                    { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
               )
               dispatch(setUser(response.data))
          } catch (e) {
               console.log(e)
          }
     }
}