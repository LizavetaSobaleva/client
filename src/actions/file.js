import { hideLoader, showLoader } from "../reducers/appReducer"
import { addFile, deleteFileAction, setFiles, setStructure } from "../reducers/fileReducer"
import { addUploadFile, changeUploadFile, showUploader } from "../reducers/uploadReducer"
import { instanceAxios } from "./user"

export function getFiles(dirId) {
  return async (dispatch) => {
    try {
      dispatch(showLoader())
      const response = await instanceAxios.get(`files${dirId ? '?parent='+dirId : ''}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      console.log(e.response.data.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function createDir(dirId, name) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.post(`files`, {
          name,
          parent: dirId,
          type: 'dir'
        },
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
        )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export function uploadFile(file, dirId) {
  return async (dispatch) => {
    try {
      const formData = new FormData()
      formData.append('file', file)
      if(dirId) {
        formData.append('parent', dirId)
      }
      const uploadFile = {name: file.name, process: 0, id: Date.now()}
      dispatch(showUploader())
      dispatch(addUploadFile(uploadFile))
      const response = await instanceAxios.post(`files/upload`, formData, {
                headers: {Authorization: `Bearer ${localStorage.getItem('token')}`},
                onUploadProgress: progressEvent => {
            const totalLength = progressEvent.event.lengthComputable ? progressEvent.event.total : progressEvent.event.target.getResponseHeader('content-length') || progressEvent.event.target.getResponseHeader('x-decompressed-content-length')
            if(totalLength) {
              uploadFile.progress = Math.round((progressEvent.loaded * 100) / totalLength)
              dispatch(changeUploadFile(uploadFile))
            }
          }
        }
      )
      dispatch(addFile(response.data))
    } catch (e) {
      alert(e.response.data.message)
    }
  }
} 

export async function downloadFile(file) {
  try {
    const response = await instanceAxios.get(`files/download?id=${file._id}`, {
      responseType: 'blob',
      headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
    )
    const blob = new Blob([response.data], { type: response.headers['content-type']})

    const downloadUrl = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = downloadUrl
    link.download = file.name

    document.body.appendChild(link)
    link.click()
    link.remove()
  } catch (e) {
    console.log(e.response.data.message)
  }
}

export function deleteFile(file) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.delete(`files?id=${file._id}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(deleteFileAction(file._id))
      alert(response.data.message)
    } catch (e) {
      alert(e.response.data.message)
    }
  }
}

export function searchFiles(search) {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/search?search=${search}`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setFiles(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}

export function getStructure() {
  return async (dispatch) => {
    try {
      const response = await instanceAxios.get(`files/structure`,
        { headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }}
      )
      dispatch(setStructure(response.data))
    } catch (e) {
      alert(e?.response?.data?.message)
    } finally {
      dispatch(hideLoader())
    }
  }
}