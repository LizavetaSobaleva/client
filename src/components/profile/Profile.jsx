import React, { useEffect, useState } from 'react'
import './profile.less'
import { useDispatch, useSelector } from 'react-redux'
import avatarLogo from '../../assets/img/user_primary.svg'
import sizeFormat from '../../utils/sizeFormat'
import SecondaryButton from '../UI/secondaryButton/SecondaryButton'
import ProgressBar from '../UI/progressBar/ProgressBar'
import { deleteAvatar, getAllUsers, uploadAvatar } from '../../actions/user'
import backIcon from "../../assets/img/back.png"
import { logout } from '../../reducers/userReducer'
import { NavLink } from 'react-router-dom'
import UsersList from './usersList/UsersList'
import Label from '../UI/label/Label'


const Profile = () => {
    const currentUser = useSelector(state => state.user.currentUser)
    const avatarUser = `${process.env.REACT_APP_API_URL}/${currentUser.avatar}`
    const avatar = currentUser.avatar ? avatarUser : avatarLogo
    const usedSpace = ((currentUser.usedSpace / currentUser.diskSpace) * 100).toFixed()
    const dispatch = useDispatch()

    const [users, setUsers] = useState([])

    const handleUploadButtonClick = () => {
        document.getElementById('avatarInput').click();
    }

    function changeHandler(e) {
        const file = e.target.files[0]
        dispatch(uploadAvatar(file))
    }
    
    const fetchUsers = async () => {
        if (currentUser.status === 'admin') {
            const fetchedUsers = await getAllUsers();
            setUsers(fetchedUsers);
        }
    };

    useEffect(() => {
        fetchUsers();
    }, [currentUser.status]);

  return (
    <>
    <NavLink to="/">
        <div className="profile__back" data-testid="profile-back">
            <img src={backIcon} alt="" className='profile__img'/>
        </div>
    </NavLink>
        <div className='profile' data-testid="profile">

            <div className="profile__avatar" data-testid="profile-avatar">
                <img src={avatar} alt="" className="profile__image"/>
                <input type="file" accept='image/*' id="avatarInput" onChange={e => changeHandler(e)} data-testid="avatar-input" />
                <label htmlFor="avatarInput">
                    {!currentUser.avatar && <SecondaryButton onClick={handleUploadButtonClick} data-testid="upload-avatar-btn">Upload avatar</SecondaryButton>}
                </label>
                {currentUser.avatar && <SecondaryButton onClick={() => dispatch(deleteAvatar())} data-testid="delete-avatar-btn">Delete avatar</SecondaryButton>}
            </div>

            <div className="profile__content" data-testid="profile-content">
                <div className="profile__name" data-testid="profile-name">{currentUser.name}</div>
                <div className="profile__status" data-testid="profile-status">
                    <Label status={currentUser.status}>{currentUser.status}</Label>
                </div>
                <div className="profile__email" data-testid="profile-email">{currentUser.email}</div>
                <div className="profile__diskSpace" data-testid="profile-disk-space">
                    Used space:
                    <div className="profile__data" data-testid="profile-used-space">
                        {sizeFormat(currentUser.usedSpace)} from {sizeFormat(currentUser.diskSpace)}
                        <ProgressBar progress={usedSpace} data-testid="progress-bar"/>
                    </div>
                </div>
            </div>
        </div>

        {currentUser.status === 'admin' && (
                <UsersList users={users} fetchUsers={fetchUsers} data-testid="usersList" />
        )}

        <div className="profile__logout" onClick={() => dispatch(logout())} data-testid="logout-btn">
            <SecondaryButton>
                Log out
            </SecondaryButton>
        </div>
    </>
    
  )
}

export default Profile