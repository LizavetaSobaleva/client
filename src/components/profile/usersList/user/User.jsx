import React, { useState } from "react";
import "./user.less";
import sizeFormat from "../../../../utils/sizeFormat";
import ProgressBar from "../../../UI/progressBar/ProgressBar";
import { useDispatch } from "react-redux";
import { changeUserStatus } from "../../../../actions/user";
import PrimaryButton from "../../../UI/primaryButton/PrimaryButton";
import SecondaryButton from "../../../UI/secondaryButton/SecondaryButton";
import Label from "../../../UI/label/Label";
import Selector from "../../../UI/selector/Selector";

const User = ({ user, fetchUsers }) => {
  const dispatch = useDispatch();
  const usedSpace = ((user.usedSpace / user.diskSpace) * 100).toFixed();
  const [status, setStatus] = useState(user.status);
  const [isEditing, setIsEditing] = useState(false);

  const handleStatusChange = (newStatus) => {
    setStatus(newStatus);
  };

  const handleSave = async () => {
    await dispatch(changeUserStatus(user._id, status));
    setIsEditing(false);
    fetchUsers();
  };

  const statusOptions = [
    { value: 'standard', label: 'standard' },
    { value: 'premium', label: 'premium' },
    { value: 'admin', label: 'admin' },
  ];

  return (
    <div className="user" data-testid="user">
      <div className="user__name" data-testid="userName">{user.name}</div>
      <div className="user__email" data-testid="userEmail">{user.email}</div>

      <div className="user__status" data-testid="userStatus">
        {isEditing ? (
          <Selector
            value={status}
            options={statusOptions}
            onChange={handleStatusChange}
            data-testid="statusSelector"
          />
        ) : (
          <Label status={user.status} data-testid="userLabel">{user.status}</Label>
        )}
      </div>

      <div className="user__size" data-testid="userSize">
        {sizeFormat(user.usedSpace)} from {sizeFormat(user.diskSpace)}
        <ProgressBar progress={usedSpace} />
      </div>

      <div className="user__actions" data-testid="userActions">
        {isEditing ? (
          <PrimaryButton onClick={handleSave} data-testid="saveButton">Save</PrimaryButton>
        ) : (
          <SecondaryButton onClick={() => setIsEditing(true)} data-testid="editButton">Edit</SecondaryButton>
        )}
      </div>
    </div>
  );
};

export default User;