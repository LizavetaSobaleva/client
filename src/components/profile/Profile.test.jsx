import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { MemoryRouter } from 'react-router-dom';
import configureStore from "redux-mock-store";
import Profile from "./Profile";
import { uploadAvatar, deleteAvatar, getAllUsers } from "../../actions/user";
import { logout } from "../../reducers/userReducer";

jest.mock("../../actions/user", () => ({
  uploadAvatar: jest.fn(() => ({ type: "SET_USER" })),
  deleteAvatar: jest.fn(() => ({ type: "SET_USER" })),
  getAllUsers: jest.fn(() => Promise.resolve([{ id: 1, name: "Test User" }])),
}));

jest.mock("../../reducers/userReducer", () => ({
  logout: jest.fn(() => ({ type: "LOGOUT" })),
}));

const mockStore = configureStore([]);

let store;

beforeEach(() => {
  store = mockStore({
    user: {
      currentUser: {
        id: "1",
        name: "Test User",
        email: "test@example.com",
        avatar: null,
        usedSpace: 52428800,
        diskSpace: 104857600,
        status: "standard",
      },
    },
  });
  store.dispatch = jest.fn();
});

const renderComponent = () =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Profile />
      </MemoryRouter>
    </Provider>
  );
describe("UsersList Component", () => {
  test("renders profile component without crashing", () => {
    const { asFragment } = renderComponent();
    expect(screen.getByTestId("profile")).toBeInTheDocument();
    expect(screen.getByTestId("profile-avatar")).toBeInTheDocument();
    expect(screen.getByTestId("profile-name")).toHaveTextContent("Test User");
    expect(screen.getByTestId("profile-status")).toHaveTextContent("standard");
    expect(screen.getByTestId("profile-email")).toHaveTextContent(
      "test@example.com"
    );
    expect(screen.getByTestId("profile-used-space")).toHaveTextContent("50.0 Mb from 100.0 Mb");
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders upload avatar button when no avatar is present", () => {
    renderComponent();

    const uploadButton = screen.getByTestId("upload-avatar-btn");
    expect(uploadButton).toBeInTheDocument();
    fireEvent.click(uploadButton);

    const fileInput = screen.getByTestId("avatar-input");
    expect(fileInput).toBeInTheDocument();
    fireEvent.change(fileInput, {
      target: {
        files: [new File(["avatar"], "avatar.png", { type: "image/png" })],
      },
    });

    expect(store.dispatch).toHaveBeenCalledWith(uploadAvatar(expect.any(File)));
  });

  test("renders delete avatar button when avatar is present", () => {
    store = mockStore({
      user: {
        currentUser: {
          id: "1",
          name: "Test User",
          email: "test@example.com",
          avatar: "avatar.png",
          usedSpace: 52428800,
          diskSpace: 104857600,
          status: "standard",
        },
      },
    });

    renderComponent();

    const deleteButton = screen.getByTestId("delete-avatar-btn");
    expect(deleteButton).toBeInTheDocument();
    fireEvent.click(deleteButton);

    expect(deleteAvatar).toHaveBeenCalledWith();
  });

  test("renders users list if current user is admin", async () => {
    store = mockStore({
      user: {
        currentUser: {
          id: "1",
          name: "Admin User",
          email: "admin@example.com",
          avatar: null,
          usedSpace: 209715200,
          diskSpace: 524288000,
          status: "admin",
        },
      },
    });

    renderComponent();

    const usersList = await screen.findByTestId("usersList");
    expect(usersList).toBeInTheDocument();
    expect(getAllUsers).toHaveBeenCalled();
  });

  test("logs out when logout button is clicked", () => {
    renderComponent();

    const logoutButton = screen.getByTestId("logout-btn");
    expect(logoutButton).toBeInTheDocument();
    fireEvent.click(logoutButton);

    expect(logout).toHaveBeenCalledWith();
  });
});
