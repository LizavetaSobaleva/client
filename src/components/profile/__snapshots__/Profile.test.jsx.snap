// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList Component renders profile component without crashing 1`] = `
<DocumentFragment>
  <a
    aria-current="page"
    class="active"
    href="/"
  >
    <div
      class="profile__back"
      data-testid="profile-back"
    >
      <img
        alt=""
        class="profile__img"
        src="test-file-stub"
      />
    </div>
  </a>
  <div
    class="profile"
    data-testid="profile"
  >
    <div
      class="profile__avatar"
      data-testid="profile-avatar"
    >
      <img
        alt=""
        class="profile__image"
        src="test-file-stub"
      />
      <input
        accept="image/*"
        data-testid="avatar-input"
        id="avatarInput"
        type="file"
      />
      <label
        for="avatarInput"
      >
        <button
          class="secondaryButton"
          data-testid="upload-avatar-btn"
        >
          Upload avatar
        </button>
      </label>
    </div>
    <div
      class="profile__content"
      data-testid="profile-content"
    >
      <div
        class="profile__name"
        data-testid="profile-name"
      >
        Test User
      </div>
      <div
        class="profile__status"
        data-testid="profile-status"
      >
        <div
          class="label standard"
          data-testid="label"
        >
          standard
        </div>
      </div>
      <div
        class="profile__email"
        data-testid="profile-email"
      >
        test@example.com
      </div>
      <div
        class="profile__diskSpace"
        data-testid="profile-disk-space"
      >
        Used space:
        <div
          class="profile__data"
          data-testid="profile-used-space"
        >
          50.0 Mb from 100.0 Mb
          <div
            class="progressBar"
            data-testid="progressBar"
          >
            <div
              class="progressBar__line"
              data-testid="progressBarLine"
            >
              <div
                class="progressBar__upload"
                data-testid="progressBarUpload"
                style="width: 50%;"
              />
            </div>
            <div
              class="progressBar__percent"
              data-testid="progressBarPercent"
              style="margin-right: 50%;"
            >
              50%
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="profile__logout"
    data-testid="logout-btn"
  >
    <button
      class="secondaryButton"
      data-testid="secondaryButton"
    >
      Log out
    </button>
  </div>
</DocumentFragment>
`;
